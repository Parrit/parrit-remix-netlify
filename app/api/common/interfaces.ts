export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  people: Person[];
}

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
  people: Person[];
  roles: Role[];
}

export interface Person {
  id: string;
  name: string;
}

export interface Role {
  id: string;
  name: string;
}

export interface PairingArrangementDTO {
  id: number; // this should be changed to string
  pairingTime: string;
  pairingHistories: PairingHistory[];
}

export interface PairingHistory {
  pairingBoardName: string;
  people: Person[];
}
