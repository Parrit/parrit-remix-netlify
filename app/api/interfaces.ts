export interface FormErrors<T> {
  server?: string;
  fields?: Record<keyof T, string | null>;
  status: 400 | 401 | 500;
}

export const isFormErrors = <T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  arg: FormErrors<T> | any
): arg is FormErrors<T> => {
  return (arg as FormErrors<T>).status > 0;
};
