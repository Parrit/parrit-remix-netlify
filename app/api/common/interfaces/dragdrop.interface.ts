export type DragType = "Person" | "Role";

export enum DropType {
  Floating = "Floating",
  PairingBoard = "PairingBoard",
  TrashBin = "TrashBin",
}

export interface DragItem {
  id: string;
  name: string;
  type: DragType;
  pairing_board_id: string;
}

export interface DropItem {
  type: DropType;
}

export interface DropTarget {
  isOver: boolean;
  canDrop: boolean;
}

export interface DragResult {}
