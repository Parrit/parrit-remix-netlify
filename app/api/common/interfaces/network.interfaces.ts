export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}
