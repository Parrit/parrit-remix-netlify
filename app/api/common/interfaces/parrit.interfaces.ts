import { DragItem } from "~/api/common/interfaces/dragdrop.interface";

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  floating: PairingBoard;
  people: Person[]; // the new cannonical location for people
  roles: Role[]; // the new cannonical location for roles
}

export const FLOATING_IDX = "-1";

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
}

export interface Person extends DragItem {
  id: string;
  name: string;
  type: "Person";
}

export interface Role extends DragItem {
  id: string;
  name: string;
  type: "Role";
}

export interface ProjectPairingSnapshot {
  id: string;
  projectId: string;
  pairingTime: string;
  pairingInstances: PairingInstance[];
}

export interface PairingInstance {
  projectId: string;
  pairingBoardName: string | null;
  people: Person[];
  pairingTime: string;
}
