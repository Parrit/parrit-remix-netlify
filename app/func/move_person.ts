import {
  Project,
  Person,
  PairingBoard,
} from "~/api/common/interfaces/parrit.interfaces";
import add_person from "./add_person";
import remove_person from "./remove_person";
import { pipe } from "ramda";

export default (project: Project, person: Person, position: PairingBoard) =>
  pipe<[Project], Project, Project>(
    remove_person(person),
    add_person(person, position)
  )(project);
