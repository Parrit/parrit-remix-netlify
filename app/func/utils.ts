/**
 * These files are tested via the functionality of other tests.
 * That's what makes them utils and not their own functions
 */

import {
  Project,
  Person,
  PairingBoard,
} from "~/api/common/interfaces/parrit.interfaces";

export const can_a_pairing_be_made = (project: Project): boolean => {
  const atLeast2Floaters = project.floating.people.length >= 2;
  const atLeast1Floater = project.floating.people.length >= 1;
  const atLeast1EmptyBoard = get_empty_pairing_board(project) !== undefined;
  const atLeast1UnpairedSticker = unpaired_sticking_people(project).length >= 1;
  return (
    (atLeast1Floater && atLeast1UnpairedSticker) ||
    (atLeast2Floaters && atLeast1EmptyBoard)
  );
};

export const unpaired_sticking_people = (project: Project): Person[] =>
  project.pairingBoards
    .filter((board) => !board.exempt)
    .filter((board) => board.people.length === 1)
    .flatMap((board) => board.people);

export const get_empty_pairing_board = (
  project: Project
): PairingBoard | undefined => {
  return project.pairingBoards.find(
    (pb) => pb.people.length === 0 && !pb.exempt
  );
};

export const all_people_in_project = (project: Project): Person[] => {
  const people: Person[] = [...project.floating.people];
  project.pairingBoards.forEach((pb) =>
    pb.people.forEach((person) => people.push(person))
  );
  return people;
};

export const find_pairing_board_by_person = (
  project: Project,
  person: Person
): PairingBoard | undefined =>
  project.pairingBoards.find((pb) => pb.people.find((p) => p.id === person.id));
