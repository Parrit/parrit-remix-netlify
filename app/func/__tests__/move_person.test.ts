import { expect } from "@jest/globals";
import {
  PROJECT_START,
  FLOATING_PERSON,
  PB_BLUE,
  RED_PERSON,
  FLOATING_PARRITS,
  PB_RED,
} from "./common";
import move_person from "../move_person";
import { can_a_pairing_be_made, get_empty_pairing_board } from "../utils";
import { FLOATING_IDX } from "~/api/common/interfaces/parrit.interfaces";

describe("move_person", () => {
  it("moving from floating to blue", () => {
    const result = move_person(PROJECT_START, FLOATING_PERSON, PB_BLUE);
    const floating = result.people.filter(
      (p) => p.pairing_board_id === FLOATING_IDX
    );
    const blueies = result.people.filter(
      (p) => p.pairing_board_id === PB_BLUE.id
    );
    expect(floating.length).toBe(0);
    expect(blueies.length).toBe(1);
    expect(result).toMatchSnapshot();
    expect(can_a_pairing_be_made(result)).toBe(false); // no floating
    expect(get_empty_pairing_board(result)).toBeUndefined();
  });

  it("moving from red to floating", () => {
    const result = move_person(PROJECT_START, RED_PERSON, FLOATING_PARRITS);
    const floating = result.people.filter(
      (p) => p.pairing_board_id === FLOATING_IDX
    );
    const reddies = result.people.filter(
      (p) => p.pairing_board_id === PB_RED.id
    );
    expect(floating.length).toBe(2);
    expect(reddies.length).toBe(0);
    expect(result).toMatchSnapshot();
    expect(can_a_pairing_be_made(result)).toBe(true); // 2 floating
  });

  it("handles copies of objects", () => {
    const result = move_person(
      PROJECT_START,
      { ...RED_PERSON },
      FLOATING_PARRITS
    );
    const floating = result.people.filter(
      (p) => p.pairing_board_id === FLOATING_IDX
    );
    const reddies = result.people.filter(
      (p) => p.pairing_board_id === PB_RED.id
    );
    expect(floating.length).toBe(2);
    expect(reddies.length).toBe(0);
    expect(result).toMatchSnapshot();
    expect(can_a_pairing_be_made(result)).toBe(true); // 2 floating
  });

  xit("does not move a person that doesn't exist in the project", () => {
    throw "not yet written";
  });
});
