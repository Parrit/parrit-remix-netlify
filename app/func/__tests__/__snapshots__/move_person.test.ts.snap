// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`move_person handles copies of objects 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
    "people": [
      {
        "id": "floater_id",
        "name": "floater_name_start",
        "pairing_board_id": "-1",
        "type": "Person",
      },
      {
        "id": "red_parrit_id",
        "name": "red_parrit",
        "pairing_board_id": "red_id",
        "type": "Person",
      },
    ],
    "roles": [],
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "red_start",
      "people": [],
      "roles": [],
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
      "people": [],
      "roles": [
        {
          "id": "role1",
          "name": "role",
          "pairing_board_id": "blue_id",
          "type": "Role",
        },
      ],
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
      "people": [
        {
          "id": "absent",
          "name": "Permanent Vacation",
          "pairing_board_id": "exempt",
          "type": "Person",
        },
      ],
      "roles": [
        {
          "id": "ooo",
          "name": "OOOROLE",
          "pairing_board_id": "exempt",
          "type": "Role",
        },
      ],
    },
  ],
}
`;

exports[`move_person moving from floating to blue 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
    "people": [],
    "roles": [],
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "red_start",
      "people": [
        {
          "id": "red_parrit_id",
          "name": "red_parrit",
          "pairing_board_id": "red_id",
          "type": "Person",
        },
      ],
      "roles": [],
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
      "people": [
        {
          "id": "floater_id",
          "name": "floater_name_start",
          "pairing_board_id": "blue_id",
          "type": "Person",
        },
      ],
      "roles": [
        {
          "id": "role1",
          "name": "role",
          "pairing_board_id": "blue_id",
          "type": "Role",
        },
      ],
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
      "people": [
        {
          "id": "absent",
          "name": "Permanent Vacation",
          "pairing_board_id": "exempt",
          "type": "Person",
        },
      ],
      "roles": [
        {
          "id": "ooo",
          "name": "OOOROLE",
          "pairing_board_id": "exempt",
          "type": "Role",
        },
      ],
    },
  ],
}
`;

exports[`move_person moving from red to floating 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
    "people": [
      {
        "id": "floater_id",
        "name": "floater_name_start",
        "pairing_board_id": "-1",
        "type": "Person",
      },
      {
        "id": "red_parrit_id",
        "name": "red_parrit",
        "pairing_board_id": "red_id",
        "type": "Person",
      },
    ],
    "roles": [],
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "red_start",
      "people": [],
      "roles": [],
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
      "people": [],
      "roles": [
        {
          "id": "role1",
          "name": "role",
          "pairing_board_id": "blue_id",
          "type": "Role",
        },
      ],
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
      "people": [
        {
          "id": "absent",
          "name": "Permanent Vacation",
          "pairing_board_id": "exempt",
          "type": "Person",
        },
      ],
      "roles": [
        {
          "id": "ooo",
          "name": "OOOROLE",
          "pairing_board_id": "exempt",
          "type": "Role",
        },
      ],
    },
  ],
}
`;
